walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,11 +13,137 @@
+    vacantSlots.removeIf(timeSlot -> isSlotInsufficient(timeSlot, request));
+  }
+
+  /** Adds start/end-bounding TimeRange objects for {@queue} */
+  private void initializeOccupiedQueue(List<TimeRange> queue) {
+    queue.add(TimeRange.fromStartDuration(TimeRange.START_OF_DAY, 0));
+    queue.add(TimeRange.fromStartDuration(TimeRange.END_OF_DAY + 1, 0));
+  }
+
+  /**
+   * Function that updates an TimeRange object at index {@currentSlotIdx} with an 
+   * updated duration if there are TimeRange objects chaining/within the current object.
+   * 03:00 |    B  06:00 |              03:00          07:00
+   * |      A      |   07:00        ->    |      A      |
+   * @param queue List of TimeRange objects, sorted by start time.
+   * @param currentSlotIdx Index of element to be modified in {@queue}
+   */
+  private void combineTimeRange(List<TimeRange> queue, int currentSlotIdx) {
+    int adjSlotIdx = currentSlotIdx + 1;
+    TimeRange currentSlot = queue.get(currentSlotIdx);
+    TimeRange adjSlot = queue.get(adjSlotIdx);
+
+    // start combining adjacent/containing events with regards to current event
+    while (currentSlot.overlaps(adjSlot)) {
+      int curSlotStart = currentSlot.start();
+      int adjSlotEnd = adjSlot.end();
+      // Combine events if it is a chained event.
+      if (!currentSlot.contains(adjSlotEnd)) {
+        TimeRange combinedSlot = TimeRange.fromStartEnd(
+          curSlotStart, adjSlotEnd, false);
+          queue.set(currentSlotIdx, combinedSlot);
+      }
+      // Get next slot in line if available
+      queue.remove(adjSlotIdx);
+      if(adjSlotIdx == queue.size()) break;
+      else adjSlot = queue.get(adjSlotIdx);
+    }
+  }
+
+
+  /**
+   * Creates a vacant slot based on the provided index of the unavailable timeslot.
+   * 
+   * // |OE1|   <vacantEvent here>   |OE2|
+   * @param occupiedQueue Combined list of TimeRange with TimeRange that is unavailable
+   * @param slotIdx Index of TimeRange Object in front of the VacantSlot in {@occupiedQueue}
+   */
+  private TimeRange createVacantSlot(List<TimeRange> occupiedQueue, int slotIdx) {
+    TimeRange occupiedSlot1 = occupiedQueue.get(slotIdx);
+    TimeRange occupiedSlot2 = occupiedQueue.get(slotIdx + 1);
+    return TimeRange.fromStartEnd(occupiedSlot1.end(), occupiedSlot2.start(), false);
+  }
+
+  public Collection<TimeRange> query(Collection<Event> eventsSource, MeetingRequest request) {
+    List<TimeRange> vacantSlots= new ArrayList<TimeRange>();
+    List<TimeRange> occupiedQueue = new ArrayList<TimeRange>();
+    // Duplicate events list to support removal of events
+    List<Event> events = new ArrayList<>(eventsSource);
+
+    if (hasNoAttendees(request)) {
+      return Collections.singletonList(TimeRange.WHOLE_DAY);
+    }
+    if (isDurationInvalid(request)) {
+      return Collections.emptyList();
+    }
+
+    // Filter entries that required attendees are not joining
+    filterEvents(events, request);
```

Comment:
I'd prefer the `.removeIf` here. It's actually more self-explanatory than the generic `filterEvents` name.
Also try to avoid needing to write comments by using more descriptive naming.

Alternatively, you can also rename the method to be more descriptive.

Reply:
The idea behind having a generic filterEvents function was to have it act as a wrapper function to support any possible further filtering in the future. However, I agree that the removeIf was a better/more descriptive way of writing it. Changed to removeIf

Reply:
The future scenario may not come at all. Name it according to the current intent.
You can always refactor it when necessary.